src/config.rs

```diff
@@ -280,3 +278,169 @@ mod tests {
+const HOME: &str = "HOME";
+
+impl Pkcs11Config {
+
+    fn get_configuration(&self) -> Result<Pkcs11Config> {
+        match std::env::var(ENVVARNAME) {
+            Ok(filename) => {
+                if filename == "internal" {
+                    // TODO
+                    //return getDefaultCryptoConfigOpts()
+                }
+                if std::path::Path::new(&filename).exists() {
+                    return parse_config_file(filename);
+                }
+                // fall out of match
+            },
+            Err(e) => match e {
+                std::env::VarError::NotPresent => (), // fall out of match
+                std::env::VarError::NotUnicode(s) => return Err(anyhow!("")),
+            },
+        };
+
+        match std::env::var(XDGCONFIGHOME) {
+            Ok(envvar) => {
+                let mut p = std::path::PathBuf::from(envvar);
+                p.push(CONFIGFILE);
+                if std::path::Path::new(&p).exists() {
+                    return parse_config_file(p.to_str().unwrap().to_string());
+                }
+                // fall out of match
+            },
+            Err(e) => match e {
+                std::env::VarError::NotPresent => (), // fall out of match
+                std::env::VarError::NotUnicode(s) => return Err(anyhow!("")),
+            },
+        };
+
+        match std::env::var(HOME) {
+            Ok(envvar) => {
+                let mut p = std::path::PathBuf::from(envvar);
+                p.push(".config");
+                p.push(CONFIGFILE);
+                if std::path::Path::new(&p).exists() {
+                    return parse_config_file(p.to_str().unwrap().to_string());
+                }
+                // fall out of match
+            },
+            Err(e) => match e {
+                std::env::VarError::NotPresent => (), // fall out of match
+                std::env::VarError::NotUnicode(s) => return Err(anyhow!("")),
+            },
+        }
+
+        let mut p = std::path::PathBuf::from("/etc");
+        p.push(CONFIGFILE);
+        if std::path::Path::new(&p).exists() {
+            return parse_config_file(p.to_str().unwrap().to_string());
+        }
+        Err(anyhow!(""))
+    }
+
+    fn get_default_crypto_config_opts(&self) -> Result<Pkcs11Config> {
+
+        let mdyaml = get_default_module_directories_yaml("".to_string())?;
+
+        let config = format!("module-directories:\n\
+                              {} \
+                              allowed-module-paths:\n\
+                              {}", mdyaml, mdyaml);
+        parse_pkcs11_config_file(config.as_bytes())
+    }
+
+    fn get_user_pkcs11_config(&self) -> Result<Pkcs11Config> {
+        let empty_config = Pkcs11Config{
+            module_directories: vec![],
+            allowed_module_paths: vec![],
+        };
+        let c = self.get_configuration()?;
+        // TODO ? can we ever get an empty pkcs11config from get-configuration?
```

Comment:
see parse_config_file()... the basically need to check if these 3rd party libraries can return an empty config. We need the "ic == nil" golang case, basically.

Reply:
get-configuration could return an option, is one idea for this

Reply:
Leaving as-is. The only place that GetUserPkcs11Config is invoked seems to be from the golang's parse-helpers `CreateCryptoConfig` and `CreateDecryptCryptoConfig`:
```CreateCryptoConfig
  pkcs11config.GetUserPkcs11Config
    getConfiguration
      parseConfigFile
```
This isn't used in the repo at the moment, it seems. And even there, it returns the error and an empty config. (So the entire chain of returns is passing along empty configs with an error.) We could just as easily handle the error in a "parse helper" if we ever make one. So there's no extra case to handle yet.